---
title: "Intro to mr.mash"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Intro to mr.mash}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,comment = "#",fig.width = 5,
                      fig.height = 3,fig.align = "center",
                      fig.cap = "&nbsp;",dpi = 120)
```

This vignette demonstrates how we use `pecotmr` package to implement mr.mash to compute weights 
matrices for performing TWAS analysis. We use simulated data of a gene ($Y$) with expression 
level in multi-conditions in $N \approx 400$ individuals. We want to first identify imputable 
genotype matrix $X_{N\times P}$ ($P\approx3000$) for gene expression level using cross validation, 
then compute weight matrix for TWAS analysis input.  


## Set up your environment

Load the `pecotmr` package.

```{r load-pkgs}
devtools::load_all("~/githubrepo/pecotmr")
```


## Load data-set

```{r}
data(multitrait_data)
attach(multitrait_data)
```


The genotype matrix $X$ is centered and standardized and filtered with MAF threshold of 
0.05, missing rate threshold of 0.05

The expression matrix $Y$ is quantile normalized, and filtered out conditions with subject
less than 100. 

```{r}
dim(X)
```

```{r}
dim(Y)
```

The mrmash_data includes mixture prior matrices and prior grids for the scenario of 
cross validation and without cross validation. `prior_matrices_cv` and
`prior_grid_cv` contains list of data for each fold. In this example, we have prepared 
the data for five-fold cross validation on the partition of the subjects. 
The `prior grid` can be computed internally if `sumstat` univariate regression results 
are provided.


```{r}
names(multitrait_data)
```

```{r}
str(prior_matrices)
```

```{r}
str(prior_grid)
```

```{r}
str(sumstat)
```

## Cross validation to determine imputability of the unit

```{r}
set.seed(999)
```

## Compute Weight Matrix using mr.mash method 

We fit mr.mash with pre-computed mixture prior matrices and grid to compute weight matrix. 
In the example below, we subset 200 SNPs for shorter runtime.


```{r}
X <- X[, 1:200]

weight_methods <- list(
    mrmash_weights = list(
    prior_data_driven_matrices=prior_matrices,
    prior_grid=prior_grid)
    )
    
weight <- twas_weights(X=X,Y=Y,weight_methods=weight_methods)
```

```{r}
str(weight)
```

When `prior_grid` is not available, summary statistics should to be provided. 

```{r}
bhat <- sumstat$Bhat
sbhat <- sumstat$Shat

weight_methods <- list(
    mrmash_weights = list(
    prior_data_driven_matrices=prior_matrices,
    bhat=bhat,
    sbhat = sbhat)
    )
    
weight <- twas_weights(X=X,Y=Y,weight_methods=weight_methods)

str(weight)
```

```{r}
detach(multitrait_data)
rm(X, multitrait_data)
```

