---
title: "SuSiE RSS with QC"
output: html_document
date: "2023-12-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

We demonstrates how we use SuSiE RSS wrapper implemented in `pecotmr` package to fine-map using summary statistics and reference LD panel data precomputed and available as resources.

To load the `pecotmr` package,

```{r load-pkgs}
library(pecotmr)
```

## Load summary statistics and corresponding LD matrix

For given region of interest, we load the LD matrix and corresponding summary statistics data, where the summary statistics data should be a data frame with the columns "chrom" and "pos":

```{r}
library(data.table)
library(dplyr)
#### Load LD_meta_file
LD_meta_file = data.frame(chrom = c("chr8","chr8"), start = c(25007602,26225312), end = c(26225312, 27515963), 
path = c(paste0("/home/aw3600/MR_KMT_analysis/test/chr8_25007602_26225312.cor.xz",",","/home/aw3600/MR_KMT_analysis/test/chr8_25007602_26225312.cor.xz.bim",sep=""),
  paste0("/home/aw3600/MR_KMT_analysis/test/chr8_26225312_27515963.cor.xz",",","/home/aw3600/MR_KMT_analysis/test/chr8_26225312_27515963.cor.xz.bim",sep="")))
#### Load region of interest
region  = data.frame(chrom = 8,start = 26220000,end = 26225400)
#### Load GWAS summary statistics
AD_GWAS_path = "/mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.8/ADGWAS2022.chr8.sumstat.tsv"
sumstats = fread(AD_GWAS_path)%>%rename("pos"="position")%>%rename("chrom"="chromosome")%>%mutate(z=beta/se)%>%rename("A1"="ref","A2"="alt")
```
Then extract the LD matrix based on the `LD_meta_file`, `region` and `sumstats`,
```{r}
LD_data = load_LD_matrix(LD_meta_file, region, sumstats)
LD_data[[1]]$LD[1:5,1:5]
```

The output LD_data is a list contains the `LD` and `variants_df`, where the `variants_df` is a data frame with the columns "chr","variants","GD(genetic distance)"," pos","A1" and "A2" in a format of bim file. The row and column names of LD are identical to the elements of the `variants` in the data frame `variants_df`. The `variants_df` will be the input of `allele_qc` as the reference panel data.

## Summary statistics QC

Match summary statistics data and `variants_df` by ("chr", "A1", "A2" and "pos"), accounting for possible strand flips and major/minor allele flips (opposite effects and zscores). We need to ensure that the format of summary statistics data with the columns "chr", "pos", "A1" and "A2". Because some end positions of one LD block are input positions of another LD block, we will keep these variants after allele flip by setting the parameters `remove_dups` as "FALSE".

```{r}
allele_fip = allele_qc(sumstats, LD_data[[1]]$variants_df, match.min.prop=0.2, remove_dups=FALSE, flip=TRUE, remove=TRUE)
head(allele_flip)
```

The output `allele_flip` will be the summary statistics after allele flip.

Because we match sumstats and `LD_data[[1]]$variants_df` by chrom and position after ``allele_qc` step, we need to use the variants of output to extract the `LD.data[[1]]$LD`.

```{r}
library(dplyr)
allele_flip = allele_flip %>% mutate(variant_allele_flip = paste(chrom,pos,A1.sumstats,A2.sumstats,sep=":"))
LD_extract = LD_data[[1]]$LD[allele_flip$variant_allele_flip,allele_flip$variant_allele_flip]
```
## run SuSiE RSS:
```{r}
library(susieR)
res = susie_rss(allele_flip$z, LD_extract)
```
