
---
title: "Enrichment analysis of molecular QTL in genetic variants associated with complex traits"
author: "Ru Feng"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{intro-to-pecotmr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette demonstrates the enrichment analysis part of the `pecotmr` package, which largely follows from `fastENLOC` (https://github.com/xqwen/fastenloc) but uses `susieR` fitted objects as input to estimate priors for use with the `coloc` package (coloc v5, aka SuSiE-coloc). The main differences are:
1) now enrichment is based on all QTL variants whether or not they are inside signal clusters;
2) Causal QTL are sampled from SuSiE single effects, not signal clusters;
3) Allow a variant to be QTL for not only multiple conditions (eg cell types) but also multiple regions (eg genes).


## Set up environment and Load data

Load the `pecotmr` package for this enrichment analysis, abd load the susieR package for data checking and visualization
```{r load-pkgs}
library(pecotmr)
library(susieR)
```

Load fine mapping results. 
```{r}
gwas_path <- "gwas_toy.rds"
qtl_path <- "QTL_toy.rds"
gwas <- readRDS(gwas_path)
qtl <- readRDS(qtl_path)
```

## Input data requirements

### Check GWAS data 
In the GWAS toy data, the file is stored as a list, with the fine mapping results already stored under the first layer.
```{r}
names(gwas[[1]])
```
Posterior inclusion probabilities of GWAS data:
```{r}
susie_plot(gwas[[1]], y = "PIP")
```

### Check QTL data 
In the QTL toy data, the file is stored as a list and is more complicated. The first layer is named as the molecular trait object ID (gene), followed by different contexts as the second layer (e.g. Mic). The variant names and fine mapping results can be found under these contexts (`variant_names` and `susie_result_trimmed`). 
```{r}
names(qtl)
names(qtl[[1]])
names(qtl[[1]][["Mic"]])
names(qtl[[1]][["Mic"]][["susie_result_trimmed"]])
```
Posterior inclusion probabilities of QTL data:
```{r}
susie_plot(qtl[[1]][["Mic"]][["susie_result_trimmed"]], y = "PIP")
```

## Perform enrichment analysis 
The input for `xqtl_enrichment_wrapper` requires the paths for GWAS and QTL fine mapping data. You need to specify the list name of SuSiE results in each dataset as `xqtl_finemapping_obj` and the list name of variant names as `xqtl_varname_obj`. For example, if SuSiE results for QTL data are under `qtl[[1]]$Mic$susie_result_trimmed`, then `xqtl_finemapping_obj` would be `c("Mic", "susie_result_trimmed")`, and `xqtl_varname_obj` would be `c("Mic", "variant_names")` since the variant names are stored directly under `Mic`. 
For GWAS data, the fine mapping results and variant names are stored under the first layer directly, so you do not need to specify `gwas_finemapping_obj` and `gwas_varname_obj`. However, if your data does not follow this structure, you may need to specify these parameters accordingly.

```{r}
enrich_res <- xqtl_enrichment_wrapper(
  gwas_files = gwas_path, xqtl_files = qtl_path,
  xqtl_finemapping_obj = c("Mic", "susie_result_trimmed"), xqtl_varname_obj = c("Mic", "variant_names")
)
print(enrich_res[[1]])
print(head(enrich_res[["unused_xqtl_variants"]][[1]]))
```
The enrichment analysis output is stored as a 2-layer list. The first layer contains the enrichment results, which will be used as priors in coloc analysis. The second layer stores the QTL variants that were not detected in the GWAS dataset.

## Session information
Here are some details about the computing environment, including the versions of R, and the R packages, used to generate these results.
```{R}
sessionInfo()
```
